Add list of post favorites to seperate route

Earlopain opened the pull request:
Well, here goes nothing. 

![image](https://user-images.githubusercontent.com/14981592/112284465-913c7a00-8c89-11eb-8065-088225fdc0c0.png)

![image](https://user-images.githubusercontent.com/14981592/112286562-ba5e0a00-8c8b-11eb-970b-f6414324911c.png)

I added pagination simply by slicing the post favorites and using PaginatedArray. Only for the slice the user records will be loaded which also means that pages will probably have a different number of entries, depending on how many user have privacy mode turned on.

EDIT: Disregard the above paragraph, fetching the users happens by joining favorites to user now

I also added the Other Favorites column because it just looked plain bad when only one column was present. I don't know the performance implications of that but I imagine it won't be that high because I include `:user_status` when fetching users.


It's interesting how the favorites are stored in the `fav_string` of the post. Storing them that way instead of having to join a table with literally 100s of million entries comes in handy now. I'm pretty sure it's only purpose is to go from one post to who favorited it without joining. Pretty usefull for `fav:` queries.

Side note: I changed the test where the original method was called but the test fails before that. Generally a whole bunch of tests fail and I was wondering if I'm doing something wrong or if it is just the way it is

Earlopain committed:
[Feature] Move post favorites to seperate path

This makes them visible to normal users but moves them to a seperate page.
This operation is expensive, let's see if there is an easier way to fetch them.

Earlopain commented on app/controllers/post_favorites_controller.rb:
```diff
@@ -0,0 +1,28 @@
+    if params.has_key? :page
+      page = params[:page].to_i || 1
+    else
+      page = 1
```
Somehow this is the only place in code where this is done which makes me think theres probably a way to remove this. The problem I had was that if no page query parameter was present it was converted to 0 instead of using the code after ||

liamwhite replied:
`[params[:page].to_i, 1].max` is a more Ruby-like way to do it (and also handles more annoying cases)

Earlopain replied:
Thank you, that looks much cleaner. That also guarantees your page will always be 1 or more. Previously you would get a 404 not found for that, but I guess if you're putting negative pages your asking for it.

I force-pushed the change

Earlopain commented on app/controllers/post_favorites_controller.rb:
```diff
@@ -0,0 +1,28 @@
+  private
+
+  def page_size
+    75
```
Not sure if `page_size` should be in a controller

kiranoot commented on app/controllers/post_favorites_controller.rb:
```diff
@@ -1,24 +1,15 @@
+      query = query.where("bit_prefs & :value != :value", {value: 2**User::BOOLEAN_ATTRIBUTES.find_index("enable_privacy_mode")}).or(query.where(favorites: {user_id: CurrentUser.id}))
+    end
+    query = query.where(favorites: {post_id: @post.id})
+    query = query.order(favorites: :desc)
```
This sort order doesn't do what you expect it does. This is sorting by "favorites".* which is all columns in the favorites table. It would be better to sort these by "users".name so that it's consistent. I don't see the utility in showing users with the most favorites first, it seems to incentivize people to make more favorites even though there is an explicit cap on favorite count.

Earlopain replied:
I thought it just orders by favorites.id which would show the most recent ones first. But sorting by name is fine too, I don' really prefer one over the other

Earlopain replied:
I added explicit ordering by favorites.id which should result in the oldest favorites being show first. That way you can't cheat your way into the first page by refaving. But if I should just order by user.name do say so.

kiranoot replied:
I think ordering by name offers the least gamification of this parameter. Doing it by favorite id has some prestige around getting there first. By name should provide a reasonable compromise between being predictable, and not rewarding specific behaviors.

Earlopain replied:
I guess nothing would prevent someone from just trying to fav everything the moment it's posted (well, except the favlimit). Sorting by most recent would probably have been for novelty only anyways

kiranoot commented on app/views/post_favorites/index.html.erb:
```diff
@@ -0,0 +1,33 @@
+<div id="c-users">
```
Change to `c-post-favorites` to reflect naming convention. Otherwise javascript for the users pages will get run on this page.

kiranoot commented on app/controllers/post_favorites_controller.rb:
```diff
@@ -0,0 +1,15 @@
+
+  def index
+    @post = Post.find(params[:post_id])
+    query = User.includes(:user_status).joins("INNER JOIN favorites ON favorites.user_id = users.id")
```
The scope on the user model for `favorites` isn't used by anything, so can be removed, leaving it as simple `has_many :favorites` and then this can be converted back into the simpler join form.

Earlopain committed:
[Feature] Add pagination to list of favorites for post

This reduces the entities which need to be loaded to render the users
to the per page size.

- Use query to get from post to favorites
- Correctly order by favorites
- Remove scope from user.favorites
- Order by user name
- Fix template namespace
